android-buddy-plugin/src/main/java/com/likethesalad/android/buddy/utils/SourceElementsIterator.kt

```diff
@@ -0,0 +1,38 @@
+package com.likethesalad.android.buddy.utils
+
+import net.bytebuddy.build.Plugin
+
+class SourceElementsIterator(
+    iterators: MutableList<out Iterator<Plugin.Engine.Source.Element>>,
+    private val excludePrefixes: Set<String>
+) : ConcatIterator<Plugin.Engine.Source.Element>(iterators) {
+    private var nextElement: Plugin.Engine.Source.Element? = null
+
+    override fun hasNext(): Boolean {
+        if (nextElement == null) {
+            while (super.hasNext()) {
+                val element = super.next()
+                if (excludePrefixes.isEmpty() || excludePrefixes.none { element.name.startsWith(it) }) {
+                    nextElement = element
+                    return true
+                }
+            }
+
+            return false
+        } else {
+            return true
+        }
+    }
+
+    override fun next(): Plugin.Engine.Source.Element {
+        hasNext()
```

Comment:
Do we need to call this here? I think all consumers will call `hasNext` before calling `next` so I'm not sure if it's needed.

Reply:
I think I stumbled on a case where it has been called directly (without has next), which lead to unexpected behaviour. Actually this call here shouldn't hurt even if hasNext() has been called before, in that case it would go into the second branch and just return true. 

So it's basically avoids unexpected behaviour in all cases

Reply:
Fair enough, I guess we shouldn't expect that everybody will play by the rules all the time 👍